src/riak_kv_ddl_compiler.erl

```diff
@@ -0,0 +1,878 @@
+%% * do we care about the name of the validation module?
+
+%% @doc
+-module(riak_kv_ddl_compiler).
+
+-include("riak_kv_index.hrl").
+-include("riak_kv_ddl.hrl").
+
+-export([
+	 make_validation/1,
+	 make_validation/2
+	]).
+
+-define(NODEBUGOUTPUT, false).
+-define(DEBUGOUTPUT,   true).
+-define(MAKEIGNORE,    true).
+-define(MAKE,          false).
+
+make_validation(#ddl{} = DDL) ->
+    make_validation(DDL, ?NODEBUGOUTPUT, "/tmp").
+
+make_validation(#ddl{} = DDL, OutputDir) ->
+    make_validation(DDL, ?DEBUGOUTPUT, OutputDir).
+
+make_validation(#ddl{} = DDL, HasDebugOutput, OutputDir) ->
+    {Module, AST} = compile(DDL, HasDebugOutput, OutputDir),
+    {ok, Module, Bin} = compile:forms(AST),
+    SpoofBeam = "/tmp/" ++ atom_to_list(Module) ++ ".beam",
+    {module, Module} = code:load_binary(Module, SpoofBeam, Bin).
+
+compile(#ddl{} = DDL, HasDebugOutput, OutputDir) ->
+    #ddl{bucket = Bucket,
+	 fields = Fields} = DDL,
+    {ModName, Attrs, N} = make_attrs(Bucket, 1),
+    {Funs, N2} = recurse_over_fns([Fields], N, 1, [], []),
+    AST = Attrs ++ Funs ++ [{eof, N2}],
+    if HasDebugOutput ->
+	    ASTFileName = filename:join([OutputDir, ModName]) ++ ".ast",
+	    write_file(ASTFileName, io_lib:format("~p", [AST]));
+       el/=se ->
+	    ok
+    end,
+    case erl_lint:module(AST) of
+ 	{ok, []} ->
+     	    if
+     		HasDebugOutput ->
+		    write_src(AST, DDL, ModName, OutputDir);
+     		el/=se ->
+		    ok
+     	    end,
+     	    _Module = {ModName, AST};
+     	Other  -> exit(Other)
+    end.
+
+write_src(AST, DDL, ModName, OutputDir) ->
+    Syntax = erl_syntax:form_list(AST),
+    Header = io_lib:format("%%% Generated Module, do NOT edit~n~n" ++
+			       "Validates the DDL~n~p~n",
+			   [DDL]),
+    Header2 = re:replace(Header, "\\n", "\\\n%%% ", [global, {return, list}]),
+    Src = erl_prettypr:format(Syntax),
+    Contents  = [Header2 ++ "\n" ++ Src],
+    SrcFileName = filename:join([OutputDir, ModName]) ++ ".erl",
+    write_file(SrcFileName, Contents).
+
+write_file(FileName, Contents) ->
+    ok = filelib:ensure_dir(FileName),
+    {ok, Fd} = file:open(FileName, [write]),
+    io:fwrite(Fd, "~s~n", [Contents]),
+    file:close(Fd).
+
+recurse_over_fns([], N, _, [], Acc2) ->
+    {lists:flatten(lists:reverse(Acc2)), N};
+recurse_over_fns([], N, M, Acc1, Acc2) ->
+    recurse_over_fns(Acc1, N, M, [], Acc2);
+recurse_over_fns([Fields | T], N, M, Acc1, Acc2) ->
+    {Funs, N2, Maps, M2} = make_funs(Fields, N, M),
+    Maps2 = {tuple, [Y || {_X, Y} <- Maps]},
+    Acc1a = case Maps of
+		[] -> Acc1;
+		_  -> [Maps2 |  Acc1]
+	    end,
+    recurse_over_fns(T, N2, M2, Acc1a, [Funs | Acc2]).
+
+make_attrs(Bucket, N) when is_binary(Bucket)    ->
+    ModName = make_module_name(Bucket),
+    {ModAttr, N1} = make_module_attr(ModName, N),
+    {ExpAttr, N2} = make_export_attr(N1),
+    {ModName, [ModAttr, ExpAttr], N2}.
+
+make_funs(Fields, N, M) ->
+    {Args, Fields2} = case Fields of
```

Comment:
This case clause was kind of confusing to me. It took me a bit to realize that List is just Fields. I would change the name of that variable to make it more explicit. Also, if there is a way to break the whole case clause into a separate function, that may make the code read a bit better. It's not clear to me as a reader why the fields get put in tuples. 

Basically I guess I'm arguing for higher levels of abstraction here and elsewhere for those not familiar with the Erlang AST. If each set of make_functions was wrapped in a logical unit like `create_field_forms` where tuples with names were created that would allow the user to better understand how the AST was being built up and learn it in smaller doses.